### ISSUE TITLE
Provide ability to prioritize for tests on the class level

### ISSUE THREAD

**Issue by NikolayPianikov** [nunit/nunit#998]
- User has a lot of tests
- The user has a list of tests that failed in the last time

User would like to check the failed tests at the beginning.

I think it is extremely useful feature. We could implement this feature by adding "--priorities" argument. It works similar like a "--testlist", but does not filter tests, changes the execution order.

**Comment by NikolayPianikov** [comment]
I've implemented this feature in the 3 steps:
1. Explore the list of test and save it to the file
2. Run the failed tests
3. Run the others
But this feature could be easy used in the MSBuild or other tools if we implemented it in the NUnit console

**Comment by CharliePoole** [comment]
I like the idea of having an interface to provide execution priorities (what to run first) for a test run. I'm not 100% convinced that a command-line option is the best way to do it, but it's worth considering.

**Comment by rprouse** [comment]
I like the idea of being able to rerun failed tests, but I think the more common scenario would be for developers to just rerun failed tests, not failed tests first then all the tests.

I would call the issue, "Add ability to rerun failed tests". When tests fail, we could write out a `failed.txt` with failed tests, one per line. This could then be used with the `--testlist` option. We could also add a `--failedonly` that would be equivilent to `--testlist=failed.txt`.

If users wanted to save the last failure, they could rename `failed.txt` so it doesn't get overwritten on the next run, or we could add a `--nofailedtext` option to prevent overwriting it.

**Comment by rprouse** [comment]
@NikolayPianikov why do you want to run all the tests after the failed tests? It seems to me that with that workflow, the developer is still blocked waiting for all the tests to run. NUnit console doesn't output the results until after all the tests are run.

My idea is more like the test adapter where you can run all tests, run failed tests or run groups of tests. The console has the first and third feature, just not the run failed tests.

**Comment by CharliePoole** [comment]
I don't think we want to do more than what @rprouse suggests for the console runner. It seems to me that optimizing the order of running tests so that the user gets feedback quicker is more of a function of long-running programs like our Gui or the adapter. The NUnit 2.6.4 Gui has the ability to run failed tests and the new one will as well.

**Comment by NikolayPianikov** [comment]
It is not only the CI scenario. For example, I made changes according to TDD, I run tests locally. And I would like to run my "Red" tests in the beginning and after that I run others "Green" tests to check a regression. I think it is a common scenario for TDD (of course we could run unit tests only for changed classes and "--where" allows it).

**Comment by CharliePoole** [comment]
@NikolayPianikov Doesn't this approach call for running only the failed tests rather than having the console run them first?
